src/charm.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Charm the service."""
+
+import logging
+import socket
+from typing import Dict, List, Optional
+
+from charms.loki_k8s.v1.loki_push_api import _PebbleLogClient
+from cosl import JujuTopology
+from loki import (
+    CONTAINER_NAME,
+    LOKI_CERT_FILE,
+    Loki,
+)
+from loki_cluster import (
+    ConfigReceivedEvent,
+    LokiClusterRequirer,
+    LokiRole,
+)
+from ops.charm import CharmBase, CollectStatusEvent
+from ops.framework import BoundEvent, Object
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class LokiWorkerK8SOperatorCharm(CharmBase):
+    """A Juju Charmed Operator for Loki."""
+
+    _loki = None
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._container = self.unit.get_container(CONTAINER_NAME)
+        self._loki = Loki(self._container, self._loki_roles, self.model)
+        self._topology = JujuTopology.from_charm(self)
+        self._loki_cluster = LokiClusterRequirer(self)
+        self._log_forwarder = ManualLogForwarder(
+            charm=self,
+            loki_endpoints=self._loki_cluster.get_loki_endpoints(),
+            refresh_events=[
+                self.on["loki-cluster"].relation_joined,
+                self.on["loki-cluster"].relation_changed,
+            ],
```

Comment:
1. Code order wise, iirc we concluded that an `update` method was more robust?
2. What about loki-cluster-departed?